single_kernel_mongo/managers/mongo.py

```diff
@@ -0,0 +1,415 @@
+        data_interface.update_relation_data(relation.id, {"data": json.dumps(new_data)})
+
+    def add_user(self, relation: Relation):
+        """Add a user for this relation."""
+        managed_users = self.state.app_peer_data.managed_users
+        data_interface = DatabaseProviderData(
+            self.model,
+            relation.name,
+        )
+        # We do nothing if the Database Requested event has not run yet.
+        if not data_interface.fetch_relation_field(relation.id, "database"):
+            logger.info(f"Database Requested for {relation} has not run yet, skipping.")
+            raise DatabaseRequestedHasNotRunYetError
+
+        username = f"relation-{relation.id}"
+        with MongoConnection(self.state.mongo_config) as mongo:
+            has_user = mongo.user_exists(username)
+
+        # We do nothing if the user already exists in DB.
+        if has_user:
+            return
+
+        with MongoConnection(self.state.mongo_config) as mongo:
+            config = self.get_config(
+                username,
+                None,  # We are creating the user, which means we don't have password for it yet
+                data_interface,
+                relation.id,
+            )
+            logger.info("Create relation user: %s on %s", config.username, config.database)
+
+            mongo.create_user(config)
+            managed_users.add(username)
+            data_interface.set_database(relation.id, config.database)
+            data_interface.set_credentials(relation.id, config.username, config.password)
+
+        self.state.app_peer_data.managed_users = managed_users
+
+    def update_user(self, relation: Relation) -> None:
+        """Add the user for this relation."""
+        data_interface = DatabaseProviderData(
+            self.model,
+            relation.name,
+        )
+
+        username = f"relation-{relation.id}"
+        password = data_interface.fetch_my_relation_field(relation.id, "password")
+        with MongoConnection(self.state.mongo_config) as mongo:
+            has_user = mongo.user_exists(username)
+        if has_user:
+            return
+        with MongoConnection(self.state.mongo_config) as mongo:
+            config = self.get_config(
+                username,
+                password,
+                data_interface,
+                relation.id,
+            )
+
+            logger.info("Update relation user: %s on %s", config.username, config.database)
+            mongo.update_user(config)
+            logger.info("Updating relation data according to diff")
+
+    def remove_user(self, relation: Relation):
+        """Removes a user from Charmed MongoDB.
+
+        Note this only removes users that this application of Charmed MongoDB is responsible for
+        managing. It won't remove:
+        1. users created from other applications
+        2. users created from other mongos routers.
```

Comment:
I'll update this comment

Reply:
Right now the config-server removes the users, here with `handle_mongos_k8s_users_removal ` you can see [the PR here](https://github.com/canonical/mongodb-k8s-operator/commit/e8b6f2dafe46fd6e6c6f5cafbea7ce65f3037d51#diff-d73e32dada64ed76c19b1c1ba19a9df7ffc1d0084f1b0c676d77ca9e92bff82eR373) - so probably fine to keep it as is for now since the functionality is fine. but I will suggest a commet to make things more obvious

Reply:
Waiting for the comment suggestion :)